decisions/DR004-Core-language-support.md

```diff
@@ -0,0 +1,476 @@
+    simpler as all core API functionality is inherently available
+    through idiomatic `C`.
+-   Inherently presents a stable ABI isolation layer between hosts and
+    plugins.
+
+#### Cons
+
+-   The data model of OpenAssetIO _requires_ `map` and `optional` types.
+    Implementing these in pure `C` either requires significant coding,
+    or would add additional third-party dependencies. Either of these
+    add significantly to code complexity, maintenance overhead and
+    project risk.
+-   C is notoriously unsafe with regards to type-safety and memory
+    management. Higher-level systems programming languages, such as C++
+    and Rust, add protective mechanisms (both compile-time and run-time)
+    with little or no performance overhead.
+-   C is less expressive than higher-level languages, missing features
+    such as inheritance and generic programming, which would likely
+    increase development time.
+-   Introduces additional boilerplate and overhead in the `C++` to `C++`
+    high-performance path.
+
+### Option 3
+
+`C++` core library, with critical business logic exposed via `C`
+interfaces. Bound to `Python`.
+
+**Estimated Cost**: Medium
+
+#### Pros
+
+-   The `C` API is still a first-class citizen within the architecture,
+    aiding future adoption in other languages and the development of
+    optional ABI isolation layers.
+-   The core implementation can leverage the `C++17` standard library,
+    and language features, dramatically reducing code complexity.
+-   The `C++` to `C++` high-performance path is natively supported.
+
+#### Cons
+
+-   Easy to inadvertently 'leak' `C++` specific practices through the
+    `C` API.
+-   The standard library can add 'bloat' when certain conveniences are
+    used.
+
+## Outcome
+
+OpenAssetIO will adopt **Option 3** and implement the API core in `C++`.
+Where this contains language-universal business logic (as opposed to
+language-specific boilerplate), it will be exposed through an idiomatic
+`C` interface that can be used to bind to arbitrary languages at some
+future point.
+
+The appendix contains a sketch illustration of how the implementation of
+a typical API component may look.
+
+## Rationale
+
+This arrangement provides the best possible intersection of the
+requirements outlined in the Relevant data section.
+
+-   The `C++` to `C++` fast path can be implemented directly with
+    minimal middleware, at the expense of ABI coupling/etc.
+-   Compiler isolation via a `C` function pointer suite, or other
+    technique such as `protobuf` can be implemented as a shim between a
+    host and a plugins implementation.
+-   Future language bindings should be more readily implementable as the
+    `C` interface is part of the core design, rather than an
+    afterthought.
+-   The core code base can make maximum use of standard library
+    containers and memory management tools, reducing entire categories
+    of programming errors.
+
```

Comment:
my previous point being support of this point, of course, but trying to lay it out explicitly